### ISSUE TITLE
DisallowAllDefaultValues behaves unexpectedly for null

### ISSUE THREAD

**Issue by Varorbc** [dotnet/runtime#83797]
> ### `RequiredAttribute.DisallowAllDefaultValues`
> The `RequiredAttribute` now allows validating that structs do not equal their default values. For example:
> 
> ```cs
> [Required(DisallowAllDefaultValues = true)]
> public Guid MyGuidValue { get; set; }
> ```
> 
> will fail validation if its value equals `Guid.Empty`.
>

If the property is null, it does not work well. Because the default value of the nullable property is null, and the `RequiredAttribute` itself is validation null, `DisallowAllDefaultValues` will meaningless. I expect validation to be successful when the value is Guid.Empty.

_Originally posted by @eiriktsarpalis in https://github.com/dotnet/core/issues/8134#issuecomment-1458216222_

**Comment by jeffhandley** [comment]
Hi @Varorbc, to make sure your issue is understood, can you confirm that the scenario you're concerned about is the following?

```c#
public class Entity_DisallowDefault
{
    [Required(DisallowAllDefaultValues = true)]
    public Guid? EntityId { get; set; }
}

[Fact]
public void DisallowDefault_Null()
{
    Entity_DisallowDefault entity = new();
    List<ValidationResult> results = new();
    ValidationContext context = new(entity);
    bool isValid = Validator.TryValidateObject(entity, context, results);

    Assert.False(isValid);
}

[Fact]
public void DisallowDefault_Empty()
{
    Entity_DisallowDefault entity = new() { EntityId = Guid.Empty };
    List<ValidationResult> results = new();
    ValidationContext context = new(entity);
    bool isValid = Validator.TryValidateObject(entity, context, results);

    Assert.True(isValid); // YOU ARE EXPECTING THIS TO BE INVALID, BUT IT IS VALID
}
```

**Comment by pinkfloydx33** [comment]
I can sort of see both sides of this. Compare it to strings with `AllowEmptyStrings=false`; The attribute itself prevents `null` and the additional argument prevents the empty string. In this case it feels like it should do something similar: the attribute's base behavior prevents `null`, and setting `DisallowAllDefaultValues=true` prevents `Guid.Empty`.

Of course nullable structs are legitimate types unlikes then difference between `string` and `string?` so it's not quite the same thing. The default value in this case _is_ `null` so the attribute is doing what it claims to. Plus, if it's meant to be required why bother making it nullable in the first place?

If this is being used in MVC, a `null`/absent Guid in the request _should_ result in `Guid.Empty` for a non-nullable property, and `DisallowAllDefaultValues` would take care of the rest. If using the STJ model binder, `JsonRequired` would result in different errors for `null` (in both scenarios) that don't even make it to data annotations.

On the flip side, keeping the property as nullable could have some value depending on your use case. For example, `null` might be a valid value whereas `Guid.Empty` isn't, like if you need to distinguish between the absent (null) and default/empty cases. In this scenario  you'd probably want to allow `null` but _not_ `Guid.Empty` but there's no combination of settings for the attribute that can achieve that. This is why we made our own `DisallowEmpty` attribute that allows `null` but fails for `Guid.Empty` and collections of size=0; we fallback to `RequiredAtrribute` for everything else.

TLDR: I think the attribute is doing the "right"--or least surprising--thing for the situation. Perhaps the current binary nature of the property needs to be changed to an enum with various knobs? Otherwise a custom validation attribute could help

**Comment by eiriktsarpalis** [comment]
I can see why you might expect this behaviour, but I think it's important to consider what `RequiredAttribute` is meant to support and what motivated the inclusion of `DisallowAllDefaultValues`: generally speaking it is meant to detect the absence of a value being set by the underlying deserializer. For unbound properties of type `Guid` you would expect a value of `default(Guid)` and for unbound properties of type `Guid?` you would expect a value of `null`. If the value is `(Guid?)default(Guid)` instead, that is almost always an indication that this value has been set explicitly by a deserializer and therefore a `RequiredAttribute` should accept it as valid.

As @pinkfloydx33 already points out, the `DisallowAllDefaultValues` property does what it says on the tin, and extending it to reject `(T?)default(T)` values would require a different name, and I would say lies beyond the scope of what `RequiredAttribute` is meant to handle.

**Comment by jeffhandley** [comment]
I'm going to reopen this. I think we should reconsider this design given the immediate feedback we've received (now from multiple people).

**Comment by eiriktsarpalis** [comment]
> I'm going to reopen this. I think we should reconsider this design given the immediate feedback we've received (now from multiple people).

I'm conflicted about this. If we did change the behaviour to include `(TStruct?)default(TStruct)`, I anticipate that an equal amount would be surprised by that behaviour in different contexts (e.g. validation failing for `int?` properties that are set to `0`). My recommendation is to _not_ make this change for the following reasons:

1. It would necessitate a name change for the property to reflect the new semantics.
2. `RequiredAttribute` works as intended for nullable types without need for the new flag, which was specifically introduced to account for non-nullable structs.
3. A property containing `(TStruct?)default(TStruct)` typically indicates that this value was explicitly set by a deserializer, and is not a indication that no value or a null value was provided.

cc @geeknoid

**Comment by jeffhandley** [comment]
Another approach on this could be to remove this functionality from `RequiredAttribute` and instead use a different validator for this logic. Too bad we can't use `DeniedValuesAttribute` for it, but something targeted for this scenario might be better anyway.
